sparrow-py/pysrc/sparrow_py/expr.py

```diff
@@ -0,0 +1,209 @@
+"""Defines classes representing Kaskada expressions."""
+
+import sys
+from typing import Callable
+from typing import Tuple
+from typing import Union
+from typing import final
+
+from sparrow_py import ffi
+
+
+class Expr(object):
+    """A Kaskada expression."""
+
+    ffi: ffi.Expr
+
+    def __init__(self, name: str, *args: "Expr") -> None:
+        """
+        Construct a new expression.
+
+        Parameters
+        ----------
+        name : str
+            Name of the operation to apply.
+        args : list[Expr]
+            List of arguments to the expression.
+
+        Raises
+        ------
+        TypeError # noqa: DAR402
+            If the argument types are invalid for the given function.
+        """
+        ffi_args = [arg.ffi for arg in args]
+        try:
+            self.ffi = ffi.Expr(name, ffi_args)
+        except TypeError as e:
+            e = TypeError(str(e))
+            if sys.version_info >= (3, 11):
+                # If we can add notes to the exception, indicate the types.
+                # This works in Python >=3.11
+                for n, arg in enumerate(args):
+                    e.add_note(f"Type of arg[{n}]: {arg.data_type()}")
+            raise e
+
+    def __repr__(self) -> str:
+        """Return an unambiguous string representing the expression."""
+        return repr(self.ffi)
+
+    def __str__(self) -> str:
+        """Return a readable string representing the expression."""
+        return str(self.ffi)
+
+    def data_type(self) -> str:
+        """Return a string describing the data type."""
+        return self.ffi.data_type_string()
+
+    @final
+    def pipe(
```

Comment:
This is currently based on Pandas. It allows:

`a.pipe(f, b...) => f(a, b...)`

and

`a.pipe((f, k), b...) => f(b..., k=a)`

It has errors if the first arguments (that use `b...`) are the ones named `k`. We can probably fix that with some code changes -- so we can decide how we want to implement this. I like this as an alternative to adding a bunch of functions (and thus names) to the `Expr` class, and also as a way of documenting the behavior of pipe.